File: Full/cpp/Full.cc
// Original author: Felice Pantaleo (CERN) <felice.pantaleo@cern.ch>

#include "TruthGraphAnalysis/Full/cpp/Full.h"

#include <cmath>
#include <map>
#include <string>
#include <vector>

#include "PhysicsTools/TruthInfo/interface/Branch.h"
#include "PhysicsTools/TruthInfo/interface/TruthLevels.h"
#include "TruthGraphAnalysis/Common/interface/ExampleSupport.h"

namespace tga::full {

  using tga::decayMode;
  using tga::firstChildWithPdgId;
  using tga::fixed;
  using tga::isLepton;
  using tga::isWeakBoson;
  using tga::lastCopy;
  using tga::membersOf;
  using tga::productionSiblings;
  using tga::signalRoots;

  void run(truth::Graph const& graph, std::ostream& out) {
    std::map<uint64_t, std::size_t> particlesOf;
    for (auto const& particle : graph.particles()) {
      ++particlesOf[particle.eventId];
    }
    for (auto const& [eventId, count] : particlesOf) {
      out << "full: interaction bx " << truth::bunchCrossingOf(eventId) << " index " << truth::eventIndexOf(eventId)
          << ": " << count << " particles\n";
    }
    for (const bool signal : {true, false}) {
      std::size_t objects = 0;
      std::size_t withoutMomentum = 0;
      double energy = 0.;
      // The members the producer stamped, as the python twin reads them.
      for (auto const& member : graph.particles()) {
        if (!member.isAtLevel(truth::LevelFlag::ReconstructableFinalState) || member.isSignal() != signal) {
          continue;
        }
        ++objects;
        energy += member.momentum.energy();
        withoutMomentum += member.hasMomentum() ? 0 : 1;
      }
      out << "full: " << (signal ? "signal" : "pileup") << " reconstructable final state " << objects << " objects, "
          << fixed(energy, 1) << " GeV, " << withoutMomentum << " without momentum\n";
    }
  }

}  // namespace tga::full

File: Full/py/full.py
#!/usr/bin/env python3
# Original author: Felice Pantaleo (CERN) <felice.pantaleo@cern.ch>

"""What does the whole event hold, interaction by interaction, and how large is the reconstructable final state of the signal and of the pile-up?

    python3 full.py ../../Common/fixtures/top.json
    python3 full.py step3.root -n 2

The C++ twin is cpp/Full.cc and prints the same lines.
"""

import argparse
import math
import sys

from PhysicsTools.TruthInfo.graphTools import bunchCrossingOf, eventIndexOf
from TruthGraphAnalysis.Common.exampleSupport import (LEPTONS, NEUTRINOS, add, decayMode, eta,
                                                      firstChildWithPdgId, graphsFrom, mass,
                                                      productionSiblings, pt)


def run(graph, out=sys.stdout):
    """Prints one line per object of interest."""
    for eventId in graph.interactions():
        out.write("full: interaction bx %d index %d: %d particles\n"
                  % (bunchCrossingOf(eventId), eventIndexOf(eventId), len(graph.particlesOfInteraction(eventId))))
    for side, keep in (("signal", graph.isSignal), ("pileup", graph.isFromPileup)):
        members = [i for i in graph.particlesOfLevel("reconstructableFinalState") if keep(i)]
        energy = sum(graph.p4(i)[3] for i in members)
        out.write("full: %s reconstructable final state %d objects, %.1f GeV, %d without momentum\n"
                  % (side, len(members), energy, sum(1 for i in members if not graph.hasMomentum(i))))


def main():
    parser = argparse.ArgumentParser(description=__doc__, formatter_class=argparse.RawDescriptionHelpFormatter)
    parser.add_argument("-n", "--maxEvents", type=int, default=-1)
    parser.add_argument("inputs", nargs="+", help="an EDM file, or one or more JSON dumps")
    args = parser.parse_args()
    for graph in graphsFrom(args.inputs, args.maxEvents):
        print("== event %s" % (graph.eventId,))
        run(graph)


if __name__ == "__main__":
    main()
